src/server/utils/file.js

```diff
@@ -0,0 +1,135 @@
+import { promises as fs } from "fs";
+import path from "path";
+import slugify from "slugify";
+import { loadFront } from "yaml-front-matter";
+
+export async function getFiles(dir, recursive = true) {
+  const dirents = await fs.readdir(dir, { withFileTypes: true });
+  const files = await Promise.all(
+    dirents.map((dirent) => {
+      const res = path.resolve(dir, dirent.name);
+
+      if (!recursive && dirent.isDirectory()) return [];
+
+      return dirent.isDirectory() ? getFiles(res) : res;
+    })
+  );
+
+  return files.flat();
+}
+
+const LINE_BREAK_REGEX = /[\r\n]+/;
+const HEADING_REGEX = /^(?<prefix>#{1,6})\s(?<name>.*)$/i;
+
+const findClosestParent = (depth, startingParent) => {
+  let parent = startingParent;
+
+  while (depth <= parent.depth) {
+    parent = parent.parent;
+  }
+
+  return parent;
+};
+
+const removeLinks = (node) => {
+  delete node.parent;
+
+  node.children.forEach(removeLinks);
+};
+
+const defaultSlugify = (str) => slugify(str, { lower: true });
+
+export const createPageTOC = async (filePath, options = {}) => {
+  const { createSlug = defaultSlugify, maxDepth = 3 } = options;
+  const md = await fs.readFile(filePath, { encoding: "utf8" });
+  const lines = md.split(LINE_BREAK_REGEX).map((line) => line.trim());
+
+  const rootParent = {
+    depth: 0,
+    children: [],
+    parent: null,
+  };
+
+  let parent = rootParent;
+
+  for (const line of lines) {
+    const match = line.match(HEADING_REGEX);
+
+    if (match == null) continue;
+
+    const { prefix = "#", name = "" } = match.groups;
+    const depth = prefix.length;
+
+    if (depth >= maxDepth) continue;
+
+    const closestParent =
+      depth > parent.depth ? parent : findClosestParent(depth, parent);
+
+    const slug = createSlug(name);
+
+    const next = {
+      name,
+      slug,
+      depth,
+      parent: closestParent,
+      children: [],
+    };
+
+    closestParent.children.push(next);
+
+    parent = next;
+  }
+
+  // Clean up circular refs
+  removeLinks(rootParent);
+
+  return rootParent.children;
+};
+
+const defaultMapPathToRoute = (dir, filePath) =>
+  path.relative(dir, filePath).replace(/\.md$/i, "");
+
+const defaultSortFn = (a, b) => {
+  if (a.parentRoute === b.parentRoute)
+    return Math.sign(a.metadata?.order - b.metadata?.order);
+
+  if (a.parentRoute.startsWith(b.parentRoute)) return 1;
+
+  return 0;
+};
+
+const defaultNameFn = (page) => page.filePath;
+
+export const getPages = async (dir, options = {}) => {
+  const {
+    mapPathToRoute = defaultMapPathToRoute,
+    sort = defaultSortFn,
+    name = defaultNameFn,
```

Comment:
It would be cool to add a `baseRoute` option that could be used in `mapPathToRoute` so that the routes could end up like `docs/topic-one/one`.

Reply:
This would be helpful if you had pages from more than one base route that needed to have links show up on the same sidebar for example

Reply:
Sure, can do!

Reply:
> @treyhoover this is looking good! Thanks for splitting out the page TOC helper. The one thing that is still missing is the directory / file based TOC helper. We could just use the list of the `route` attributes you add to each page, and format that into a TOC on the client side though.

That's what I was thinking originally. Do you think we could tackle the file-based helper separately if we decide we need it? Otherwise I can start on it in this PR but there are a lot of cases to consider there that I think may be more of a client concern.

Reply:
Yeah, I'm fine leaving the TOC for the client side, especially with the baseRoute option added.

Reply:
That way the sidebar Paula writes can just expect an array of route strings, and she can handle the hierarchy and linking for them no matter how many different types of routes it includes. `/docs/` , `/recipes/` etc.